File: sdk/python/src/securenotify/managers/publish_manager.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2026 KirkyX. All rights reserved.

from typing import Optional, Dict, Any
from datetime import datetime

from securenotify.types.api import (
    MessagePublishRequest,
    MessagePublishResponse,
    MessagePriority,
    QueueStatusInfo,
)
from .base import BaseManager
from securenotify.utils.http import validate_channel_id


class PublishManager(BaseManager):
    async def send(
        self,
        channel: str,
        message: str,
        priority: MessagePriority = MessagePriority.NORMAL,
        sender: Optional[str] = None,
        encrypted: bool = True,
        signature: Optional[str] = None,
        cache: bool = True,
    ) -> MessagePublishResponse:
        """Send a message to a channel.

        Raises:
            ValueError: If channel or message is empty or invalid.
            SecureNotifyApiError: On API error.
        """
        # Validate channel ID format (SECURITY FIX)
        if not validate_channel_id(channel):
            raise ValueError(
                f"Invalid channel ID '{channel}'. "
                "Channel ID must be 1-256 characters and contain only alphanumeric characters, hyphens, and underscores."
            )

        if not message or not isinstance(message, str):
            raise ValueError("Message must be a non-empty string")

        request = MessagePublishRequest(
            channel=channel,
            message=message,
            priority=priority,
            sender=sender,
            encrypted=encrypted,
            signature=signature,
            cache=cache,
        )
        return await self._execute("publish_message", request)

    async def get_queue_status(self, channel: str) -> QueueStatusInfo:
        data = await self._execute("get_queue_status", channel)
        return QueueStatusInfo(
            channel=data["channel"],
            pending_count=data["pending_count"],
            priority_counts=data.get("priority_counts", {}),
        )

    async def send_critical(
        self,
        channel: str,
        message: str,
        sender: Optional[str] = None,
        encrypted: bool = True,
    ) -> MessagePublishResponse:
        return await self.send(
            channel=channel,
            message=message,
            priority=MessagePriority.CRITICAL,
            sender=sender,
            encrypted=encrypted,
        )

    async def send_high(
        self,
        channel: str,
        message: str,
        sender: Optional[str] = None,
        encrypted: bool = True,
    ) -> MessagePublishResponse:
        return await self.send(
            channel=channel,
            message=message,
            priority=MessagePriority.HIGH,
            sender=sender,
            encrypted=encrypted,
        )

    async def send_bulk(self, channel: str, message: str) -> MessagePublishResponse:
        """Send a bulk priority message.

        For low-priority messages that can be delayed.
        """
        return await self.send(
            channel=channel, message=message, priority=MessagePriority.BULK
        )

File: sdk/java/src/main/java/securenotify/managers/PublishManager.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 KirkyX. All rights reserved.

package securenotify.managers;

import securenotify.types.ApiRequest;
import securenotify.types.ApiResponse;
import securenotify.types.MessageInfo;
import securenotify.utils.HttpClient;
import securenotify.utils.RetryHandler;

import java.util.Map;

/**
 * Manager for message publishing operations.
 */
public class PublishManager {

    private final HttpClient httpClient;
    private final RetryHandler retryHandler;

    public PublishManager(HttpClient httpClient) {
        this(httpClient, RetryHandler.DEFAULT);
    }

    public PublishManager(HttpClient httpClient, RetryHandler retryHandler) {
        this.httpClient = httpClient;
        this.retryHandler = retryHandler;
    }

    /**
     * @param priority  Message priority (critical, high, normal, low, bulk)
     */
    public MessageInfo.MessagePublishResponse send(String channel, String message, String priority,
                                                    String sender, Boolean cache, Boolean encrypted,
                                                    Boolean autoCreate, String signature) throws Exception {
        ApiRequest<?> request = ApiRequest.builder()
                .channel(channel)
                .message(message)
                .priority(priority)
                .sender(sender)
                .cache(cache != null ? cache : true)
                .encrypted(encrypted)
                .autoCreate(autoCreate != null ? autoCreate : true)
                .signature(signature)
                .build();

        return retryHandler.execute(() -> {
            try {
                return httpClient.post("api/publish", request, MessageInfo.MessagePublishResponse.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).getData();
    }

    public MessageInfo.MessagePublishResponse send(String channel, String message) throws Exception {
        return send(channel, message, "normal", null, true, false, true, null);
    }

    public MessageInfo.MessagePublishResponse sendCritical(String channel, String message) throws Exception {
        return send(channel, message, "critical", null, true, false, true, null);
    }

    public MessageInfo.MessagePublishResponse send(String channel, String message, String sender) throws Exception {
        return send(channel, message, "normal", sender, true, false, true, null);
    }

    public MessageInfo.QueueStatusResponse getQueueStatus(String channel) throws Exception {
        return retryHandler.execute(() -> {
            try {
                return httpClient.get("api/publish", Map.of("channel", channel), MessageInfo.QueueStatusResponse.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).getData();
    }

    public MessageInfo getMessage(String messageId) throws Exception {
        return retryHandler.execute(() -> {
            try {
                return httpClient.get("api/publish/" + messageId, null, MessageInfo.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).getData();
    }
}
